File: AspNetMvcUnleashed/0672329980_source/Chapter14/CS/MvcApplication1/MvcApplication1/Controllers/DownLinkController.cs
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using MvcApplication1.Models;

namespace MvcApplication1.Controllers
{

    public class ProductsVDM
    {
        public ProductsVDM(IEnumerable<Category> categories, IEnumerable<Product> products)
        {
            this.Categories = categories;
            this.Products = products;
        }

        public IEnumerable<Category> Categories { get; set; }
        public IEnumerable<Product> Products { get; set; }
    }

    public class DownLinkController : Controller
    {
        private ProductsDBEntities _entities = new ProductsDBEntities();

        public ActionResult Index()
        {
            var categories = _entities.CategorySet.ToList();
            var products = new List<Product>();
            return View(new ProductsVDM(categories, products));
        }


        public ActionResult Details(int id)
        {
            var products = from p in _entities.ProductSet
                           where p.CategoryId == id
                           select p;

            if (Request.IsAjaxRequest())
            {
                return PartialView("Details", products);
            }
            else
            {
                var categories = _entities.CategorySet.ToList();
                return View("Index", new ProductsVDM(categories, products));
            }

        }


    }
}

File: AspNetMvcUnleashed/0672329980_source/Chapter14/VB/MvcApplication1/MvcApplication1/Controllers/DownLinkController.vb
Public Class ProductsVDM

    Private _categories As IEnumerable(Of Category)
    Private _products As IEnumerable(Of Product)

    Public Sub New(ByVal categories As IEnumerable(Of Category), ByVal products As IEnumerable(Of Product))
        _categories = categories
        _products = products
    End Sub

    Public ReadOnly Property Categories() As IEnumerable(Of Category)
        Get
            Return _categories
        End Get
    End Property

    Public ReadOnly Property Products() As IEnumerable(Of Product)
        Get
            Return _products
        End Get
    End Property
End Class

Public Class DownLinkController
    Inherits Controller

    Private _entities As New ProductsDBEntities()

    Public Function Index() As ActionResult
        Dim categories = _entities.CategorySet.ToList()
        Dim products = New List(Of Product)()
        Return View(New ProductsVDM(categories, products))
    End Function


    Public Function Details(ByVal id As Integer) As ActionResult
        Dim products = From p In _entities.ProductSet _
                       Where p.CategoryId = id _
                       Select p

        If Request.IsAjaxRequest() Then
            Return PartialView("Details", products)
        Else
            Dim categories = _entities.CategorySet.ToList()
            Return View("Index", New ProductsVDM(categories, products))
        End If

    End Function


End Class
